src/data/nat/factorial.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2018 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Chris Hughes, Floris van Doorn
+-/
+import data.nat.basic
+/-!
+# The factorial function
+
+-/
+
+namespace nat
+variables {m n : ℕ}
+
+/-- `nat.factorial n` is the factorial of `n`. -/
+@[simp] def factorial : nat → nat
+| 0        := 1
+| (succ n) := succ n * factorial n
+
+localized "notation n `!`:10000 := nat.factorial n" in factorial
```

Comment:
```suggestion
localized "local notation n `!`:10000 := nat.factorial n" in nat
```
I don't think anything else would go in the factorial locale, so then there wouldn't be any benefit over
```
local notation n `!` ...
```